### PR TITLE
retry when checking output of kubectl describe e2e test

### LINKED ISSUE #41898: [k8s.io] Kubectl client [k8s.io] Kubectl describe should check if kubectl describe prints relevant information for rc and pods [Conformance]
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/41821/pull-kubernetes-e2e-gce-gci/18767/

```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:797
Feb 22 06:52:17.446: Failed to find 1 current in Replicas:	0 current / 1 desired
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:1567
```

replica set controller maybe?

@kargakis

### PR DESCRIPTION
**Pull request by mengqiy**
Add retry for kubectl describe e2e test according to https://github.com/kubernetes/kubernetes/issues/41898#issuecomment-294175426

fixes: #41898 

cc: @fejta

### CODE REVIEW

File: test/e2e/kubectl.go [hunk 1]
```diff
@@ -1630,22 +1629,43 @@ var _ = framework.KubeDescribe("Kubectl client", func() {
 })
 
 // Checks whether the output split by line contains the required elements.
-func checkOutput(output string, required [][]string) {
+func checkOutputReturnErrorString(output string, required [][]string) string {
```
Comment (kargakis) [inline test/e2e/kubectl.go]: Return error instead of string

Comment (mengqiy) [inline test/e2e/kubectl.go]: Done

**Comment by mengqiy** [comment]
@kargakis PTAL

**Comment by mengqiy** [comment]
@k8s-bot bazel test this

File: test/e2e/kubectl.go [hunk 2]
```diff
@@ -1630,22 +1629,42 @@ var _ = framework.KubeDescribe("Kubectl client", func() {
...
+func checkOutputReturnError(output string, required [][]string) error {
 	outputLines := strings.Split(output, "\n")
 	currentLine := 0
 	for _, requirement := range required {
 		for currentLine < len(outputLines) && !strings.Contains(outputLines[currentLine], requirement[0]) {
 			currentLine++
 		}
 		if currentLine == len(outputLines) {
-			framework.Failf("Failed to find %s in %s", requirement[0], output)
+			return fmt.Errorf("Failed to find %s in %s", requirement[0], output)
```
Comment (kargakis) [inline test/e2e/kubectl.go]: As a go convention, messages in fmt.Errorf start lowercased

File: test/e2e/kubectl.go [hunk 3]
```diff
@@ -1630,22 +1629,42 @@ var _ = framework.KubeDescribe("Kubectl client", func() {
...
 			}
 		}
 	}
+	return nil
+}
+
+func checkOutput(output string, required [][]string) {
+	err := checkOutputReturnError(output, required)
+	if err != nil {
+		framework.Failf("%s", err)
```
Comment (kargakis) [inline test/e2e/kubectl.go]: %v

File: test/e2e/kubectl.go [hunk 4]
```diff
@@ -1630,22 +1629,42 @@ var _ = framework.KubeDescribe("Kubectl client", func() {
...
+	}
+}
+
+func checkKubectlOutputWithRetry(required [][]string, args ...string) {
+	var err error
+	for i := 0; i < 5; i++ {
+		output := framework.RunKubectlOrDie(args...)
+		err = checkOutputReturnError(output, required)
+		if err != nil {
+			framework.Failf("%s", err)
```
Comment (kargakis) [inline test/e2e/kubectl.go]: %v

**Comment by kargakis** [comment]
Some nits, lgtm in general

This is not going to fix #41898 though

**Comment by mengqiy** [comment]
> This is not going to fix #41898 though

For the upgrade tests, this is not fixing.

Fixing upgrade tests needs to change describer code and kubectl e2e test code in release-1.4 and release-1.5 branch.
Or maybe even simpler, disable this test when kubectl version is 1.6 or newer.

File: test/e2e/kubectl.go [hunk 5]
```diff
@@ -1630,22 +1629,42 @@ var _ = framework.KubeDescribe("Kubectl client", func() {
...
+
+func checkOutput(output string, required [][]string) {
+	err := checkOutputReturnError(output, required)
+	if err != nil {
+		framework.Failf("%v", err)
+	}
+}
+
+func checkKubectlOutputWithRetry(required [][]string, args ...string) {
+	var err error
```
Comment (kargakis) [inline test/e2e/kubectl.go]: No need to declare this here since Failf will exit the test?

Comment (mengqiy) [inline test/e2e/kubectl.go]: Done

File: test/e2e/kubectl.go [hunk 6]
```diff
@@ -1630,22 +1629,41 @@ var _ = framework.KubeDescribe("Kubectl client", func() {
...
+		framework.Failf("%v", err)
+	}
+}
+
+func checkKubectlOutputWithRetry(required [][]string, args ...string) {
+	for i := 0; i < 5; i++ {
+		output := framework.RunKubectlOrDie(args...)
+		err := checkOutputReturnError(output, required)
+		if err != nil {
+			framework.Failf("%v", err)
```
Comment (kargakis) [inline test/e2e/kubectl.go]: Actually we are not supposed to fail here, right? Also this loop needs to be a PollImmediate or something

Comment (mengqiy) [inline test/e2e/kubectl.go]: I got your point. We should use Errorf and PollImmediate here.
What should the polling interval be?

Comment (mengqiy) [inline test/e2e/kubectl.go]: And the timeout?

Comment (kargakis) [inline test/e2e/kubectl.go]: Interval=1s, timeout=1m

Comment (mengqiy) [inline test/e2e/kubectl.go]: Done

**Comment by mengqiy** [comment]
@kargakis PTAL

File: test/e2e/kubectl.go [hunk 7]
```diff
@@ -1630,22 +1629,45 @@ var _ = framework.KubeDescribe("Kubectl client", func() {
...
+		framework.Failf("%v", err)
+	}
+}
+
+func checkKubectlOutputWithRetry(required [][]string, args ...string) {
+	err := wait.PollImmediate(time.Second, time.Minute, func() (bool, error) {
+		output := framework.RunKubectlOrDie(args...)
+		err := checkOutputReturnError(output, required)
+		if err != nil {
+			return false, nil
```
Comment (kargakis) [inline test/e2e/kubectl.go]: log this error

File: test/e2e/kubectl.go [hunk 8]
```diff
@@ -1630,22 +1629,45 @@ var _ = framework.KubeDescribe("Kubectl client", func() {
...
+		}
+		return true, nil
+	})
+	if err != nil {
```
Comment (kargakis) [inline test/e2e/kubectl.go]: The only error that comes back from PollImmediate is a timeout error which is far from helpful. I would suggest passing through the original error.
```go
+ var pollErr error
+	err := wait.PollImmediate(time.Second, time.Minute, func() (bool, error) {
+		output := framework.RunKubectlOrDie(args...)
+		err := checkOutputReturnError(output, required)
+		if err != nil {
+           // log the error
+           pollErr = err
+			return false, nil
+		}
+       pollErr = nil
+		return true, nil
+	})
+ if pollErr != nil {
+       framework.Failf("%v", pollErr)
+ }
```

**Comment by kargakis** [comment]
@mengqiy I am ok with this fix but we should really move this into a unit test (and I think we already have unit tests for `describe` so this test is probably redundant).

**Comment by mengqiy** [comment]
@kargakis  Agree. @pwittrock and I plan to move some of the kubectl e2e tests to a new integration test framework for kubectl. But it's not for Q2. We may do that in Q3.

**Comment by mengqiy** [comment]
@k8s-bot bazel test this

**Comment by mengqiy** [comment]
@k8s-bot cvm gce e2e test this

**Pull request merged**